pkg/remote/trans/nphttp2/server_handler.go

```diff
@@ -122,136 +122,148 @@ func (t *svrTransHandler) OnRead(ctx context.Context, conn net.Conn) error {
-
-			if methodInfo == nil {
-				unknownServiceHandlerFunc := t.opt.GRPCUnknownServiceHandler
-				if unknownServiceHandlerFunc != nil {
-					rpcinfo.Record(rCtx, ri, stats.ServerHandleStart, nil)
-					err = unknownServiceHandlerFunc(rCtx, methodName, st)
-					if err != nil {
-						err = kerrors.ErrBiz.WithCause(err)
-					}
-				} else {
-					if svcInfo == nil {
-						err = remote.NewTransErrorWithMsg(remote.UnknownService, fmt.Sprintf("unknown service %s", serviceName))
-					} else {
-						err = remote.NewTransErrorWithMsg(remote.UnknownMethod, fmt.Sprintf("unknown method %s", methodName))
-					}
-				}
+	var serviceName string
+	idx := strings.LastIndex(sm[:pos], ".")
+	if idx == -1 {
+		ink.SetPackageName("")
+		serviceName = sm[0:pos]
+	} else {
+		ink.SetPackageName(sm[:idx])
+		serviceName = sm[idx+1 : pos]
+	}
+	ink.SetServiceName(serviceName)
+
+	// set grpc transport flag before execute metahandler
+	rpcinfo.AsMutableRPCConfig(ri.Config()).SetTransportProtocol(transport.GRPC)
+	var err error
+	for _, shdlr := range t.opt.StreamingMetaHandlers {
+		rCtx, err = shdlr.OnReadStream(rCtx)
+		if err != nil {
+			tr.WriteStatus(s, convertStatus(err))
+			return
+		}
+	}
+	rCtx = t.startTracer(rCtx, ri)
+	defer func() {
+		panicErr := recover()
+		if panicErr != nil {
+			if conn != nil {
+				klog.CtxErrorf(rCtx, "KITEX: gRPC panic happened, close conn, remoteAddress=%s, error=%s\nstack=%s", conn.RemoteAddr(), panicErr, string(debug.Stack()))
 			} else {
-				if streaming.UnaryCompatibleMiddleware(methodInfo.StreamingMode(), t.opt.CompatibleMiddlewareForUnary) {
-					// making streaming unary APIs capable of using the same server middleware as non-streaming APIs
-					// note: rawStream skips recv/send middleware for unary API requests to avoid confusion
-					err = invokeStreamUnaryHandler(rCtx, rawStream, methodInfo, t.inkHdlFunc, ri)
-				} else {
-					err = t.inkHdlFunc(rCtx, &streaming.Args{Stream: st}, nil)
-				}
+				klog.CtxErrorf(rCtx, "KITEX: gRPC panic happened, error=%v\nstack=%s", panicErr, string(debug.Stack()))
 			}
+		}
+		t.finishTracer(rCtx, ri, err, panicErr)
+	}()
+
+	// set recv grpc compressor at server to decode the pack from client
+	remote.SetRecvCompressor(ri, s.RecvCompress())
+	// set send grpc compressor at server to encode reply pack
+	remote.SetSendCompressor(ri, s.SendCompress())
+
+	svcInfo := t.svcSearcher.SearchService(serviceName, methodName, true)
+	var methodInfo serviceinfo.MethodInfo
+	if svcInfo != nil {
+		methodInfo = svcInfo.MethodInfo(methodName)
+	}
 
+	rawStream := &stream{
+		ctx:     rCtx,
+		svcInfo: svcInfo,
+		conn:    newServerConn(tr, s),
+		handler: t,
+	}
+	st := newStreamWithMiddleware(rawStream, t.opt.RecvEndpoint, t.opt.SendEndpoint)
+
+	// bind stream into ctx, in order to let user set header and trailer by provided api in meta_api.go
+	rCtx = streaming.NewCtxWithStream(rCtx, st)
+	// make streaming.GetStream working for rawStream.Context()
+	rawStream.ctx = rCtx
```

Comment:
Once assigned, the bottom ```stream.Context()``` allows ```GetServerConn``` to be used to get the ```GRPCConn```, and subsequent wrapped streams won't fail as long as they don't overwrite this ctx